### ISSUE TITLE
[Q] : Why are dict values in the state converted to StateProxy classes?

### ISSUE THREAD

**Issue by thondeboer** [streamsync-cloud/streamsync#125]
I have a need to store a dict in the state, but doing that makes the dict a special calss called StateProxy, that does not behave as a dict, so I cannot continue using it in my code..

Here's my code and state[_dss'] is the problem...Can I not use state to store dicts for my own use?

```python
def analyze(state: Dict[str, Any]):
    """Analyze the data.

    Args:
        state (dict): The state dictionary.

    This function is called when the user clicks the 'Analyze' button.
    It will run DeSeq2 creating the dataframes with the results of the
    differential expression analysis.
    """
    pd.set_option('display.max_colwidth', None)
    deseq = py_DESeq2(
        design_matrix=state["_condition_sheet"], design_formula=state['comparison'])
    deseq.run_deseq()
    # Get all the comparisons with the condition values from DeSeq2
    conditions = state["condition_counts"].Condition.unique()
    dss = {}
    for a, b in combinations(conditions, 2):
        logger.warning(f'Running DESeq2 for {a} vs {b}')
        deseq.get_deseq_result(contrast=["Condition", a, b])
        dss[f"{a}_v_{b}"] = deseq.deseq_result
    logger.warning(f'Got these comparisons: {list(dss.keys())}')
    logger.warning(f"Type dss: {type(dss)}")
    state["_dss"] = dss
    logger.warning(f'Type state[dss]: {type(state["_dss"])}')

    state["_gene_annotation"] = deseq.get_gene_annotation()
    logger.warning(f'Gene-level gene annotation: {state["_gene_annotation"] }')

    state["_gene_abundance"] = deseq.get_gene_abundance()
    logger.warning(f'Gene-level gene abundance: {state["_gene_abundance"]}')

    state.set_page("analyze")```

**Comment by ramedina86** [comment]
This is a great point. To answer the question in the title, the reason why dicts are converted to StateProxy is to detect mutations, so when you do `state["my_dict"]["a"]` the parent (my_dict) gets marked as mutated.

As a workaround, you can try accessing the `state` property of the StateProxy, which contains a dictionary that stores the values. For example, `state["my_dict"].state`.

This needs to be improved somehow. I'll look into implementing `StateProxy` as an extension of `Dict` so that it behaves as a proper dictionary when needed but still being able to detect mutations.

**Comment by thondeboer** [comment]
Thanks for the explantion.I worked around the problem by making my dict a "named list"; a list of tuples, fist being the key, second being the value I needed), but i may try your state solution

**Comment by ramedina86** [comment]
@mmikita95 Mikita, given that you're deep into `StateProxy`, I thought it was a good chance to look at this issue. My idea was to make `StateProxy` inherit from `Dict` so that it'd stop "kinda behaving like a `Dict`" and become a true `Dict` (with extra capabilities for mutation detection of course). I'm open to other ideas, but I think that could work.

**Comment by ramedina86** [comment]
@mmikita95, @FabienArcellier will be working on an enhancement that solves this issue

**Comment by ramedina86** [comment]
The new State Schema feature, shipped in 0.4.0, addresses this limitation.

https://www.streamsync.cloud/state-schema.html
